src/main/java/com/cloudbees/hudson/plugins/folder/relocate/StandardHandler.java

```diff
@@ -88,42 +139,44 @@ public HandlingMode applicability(Item item) {
                 if (!permitted(item, itemGroup)) {
                     continue;
                 }
-                ItemGroup<?> p = itemGroup;
-                if (p instanceof Item) {
-                    Item i = (Item) p;
-                    // Cannot move a folder into itself or a descendant
-                    if (i == item) {
-                        continue ITEM;
-                    }
-                    // By default the move is a no-op in case you hit it by mistake
-                    if (item.getParent() == i) {
-                        result.add(itemGroup);
+                // Cannot move a folder into itself or a descendant
+                if (g == item) {
+                    continue;
+                }
+                // By default the move is a no-op in case you hit it by mistake
+                if (item.getParent() == g) {
+                    result.add(itemGroup);
+                }
+                // Cannot move an item into a Folder if there is already an item with the same name
+                if (g instanceof Folder) {
+                    // NOTE: test for Folder not AbstractFolder as Folder is mutable by users
+                    Folder folder = (Folder) g;
+                    if (folder.getItem(item.getName()) != null) {
+                        continue;
                     }
-                    // Cannot move an item into a Folder if there is already an item with the same name
-                    if (i instanceof Folder) {
```

Comment:
This was the infinite loop, namely assuming that a `Folder` would only ever have parents that were also `Folder` or `Jenkins`